File: src/algorithms/ewma.ts
export class EWMA {
	private alpha: number;
	private mean: number = 0;
	private variance: number = 0;
	private initialized: boolean = false;

	constructor(options: { halfLife: number } | { alpha: number }) {
		if ("halfLife" in options) {
			this.alpha = 1 - Math.exp(-Math.LN2 / options.halfLife);
		} else {
			this.alpha = options.alpha;
		}
	}

	update(sample: number): number {
		if (!this.initialized) {
			this.mean = sample;
			this.variance = 0;
			this.initialized = true;
		} else {
			const diff = sample - this.mean;
			const incr = this.alpha * diff;
			this.mean += incr;
			
			// EWMVar update: (1 - alpha) * (Var_prev + alpha * diff^2)
			// Or approximate incremental: Var_new = (1-alpha)*Var_old + alpha * (diff * (sample - new_mean))
			// Standard EWMVar:
			this.variance = (1 - this.alpha) * (this.variance + this.alpha * diff * diff);
		}
		return this.mean;
	}

	getValue(): number {
		return this.mean;
	}

	getStdDev(): number {
		return Math.sqrt(this.variance);
	}

	/**
	 * Returns the number of standard deviations the sample is from the mean
	 */
	getZScore(sample: number): number {
		const std = this.getStdDev();
		if (std === 0) return 0;
		return (sample - this.mean) / std;
	}

	reset(): void {
		this.mean = 0;
		this.variance = 0;
		this.initialized = false;
	}
}

File: via-core/crates/via-core/src/algo/ewma.rs
use serde::{Deserialize, Serialize};

#[derive(Serialize, Deserialize, Clone, Debug)]
pub struct EWMA {
    alpha: f64,
    mean: f64,
    variance: f64,
    initialized: bool,
}

impl EWMA {
    pub fn new(half_life: f64) -> Self {
        let alpha = 1.0 - (-std::f64::consts::LN_2 / half_life).exp();
        Self {
            alpha,
            mean: 0.0,
            variance: 0.0,
            initialized: false,
        }
    }

    pub fn update(&mut self, sample: f64) -> f64 {
        if !self.initialized {
            self.mean = sample;
            self.variance = 0.0;
            self.initialized = true;
        } else {
            let diff = sample - self.mean;
            let incr = self.alpha * diff;
            self.mean += incr;
            // Standard EWMVar update
            self.variance = (1.0 - self.alpha) * (self.variance + self.alpha * diff * diff);
        }
        self.mean
    }

    pub fn get_value(&self) -> f64 {
        self.mean
    }

    pub fn value(&self) -> f64 {
        self.mean
    }

    pub fn get_std_dev(&self) -> f64 {
        self.variance.sqrt()
    }
}
